crypto/bls12381_utils.c

```diff
@@ -1,852 +1,1173 @@
-// +build relic
-
 // this file contains utility functions for the curve BLS 12-381
-// these tools are shared by the BLS signature scheme, the BLS based threshold signature
-// and the BLS distributed key generation protocols
+// these tools are shared by the BLS signature scheme, the BLS based threshold
+// signature, BLS-SPoCK and the BLS distributed key generation protocols
 
 #include "bls12381_utils.h"
-#include "bls_include.h"
 #include "assert.h"
+#include "bls_include.h"
 
-// The functions are tested for ALLOC=AUTO (not for ALLOC=DYNAMIC)
+// compile all blst C src along with this file
+#include "blst_src.c"
+
+// make sure flow crypto types are consistent with BLST types
+void types_sanity(void) {
```

Comment:
Out of curiosity, why do we introduce own types instead of using BLST types?

Reply:
Good question. Two main reasons:
- three of these types need to be exported into the Go layer. Cgo has some constraints and some type definitions cannot be seen from Go (if they are based on macros). Therefore, it's necessary to define other exportable `struct` types (as mentioned in the comments [here](https://github.com/onflow/flow-go/blob/feature/blst-based-crypto/crypto/blst_include.h)). 
- the flow crypto types serve as an abstraction from the external low layer. If we were to replace BLST in the future, the primitives/protocols implementations (the orange box in the screenshot below - note that green is Go, blue is C) will most likely remain unchanged. Some new type names are also more descriptive of how Flow crypto is using those types.

<img width="796" alt="image" src="https://github.com/onflow/flow-go/assets/50252200/644f6da1-5602-47fd-81af-07b891d21ebc">

Reply:
Btw the sanity check function makes sure many implicit pointer casts work as expected in the low layer. Pointer casts are preferred as they are faster than conversion functions.